<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Haplotype Maps • lagniappe</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Haplotype Maps">
<meta property="og:description" content="lagniappe">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">lagniappe</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/HapMap.html">Haplotype Maps</a>
    </li>
    <li>
      <a href="../articles/MappingAndTidyverse.html">TidyversePractice</a>
    </li>
    <li>
      <a href="../articles/OysterMap.html">Processing Oyster Sites</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Haplotype Maps</h1>
                        <h4 data-toc-skip class="author">April
Wright</h4>
            
            <h4 data-toc-skip class="date">2024-03-19</h4>
      
      
      <div class="hidden name"><code>HapMap.Rmd</code></div>

    </div>

    
    
<p>This morning, we looked at cleaning and mapping spatial data. This
afternoon, we will add a genetic component to the question. These data
also come from the lab of Morgan Kelly, and concern genetic variation
among oysters.</p>
<p>One thing many scientists are interested in is haplotype maps. A
haplotype refers to a physical grouping of genetic variants that tend to
be inherited together. Usually these will be unique combinations of
variants that occur near one another on a specific chromosome. A
haplotype map does not refer to a geographic mapping, but a “map” in
which researchers can understand how a haplotypes relate to one another,
and even how they might relate to physical traits.</p>
<p>The data we are working with has been collected by Dr. Kelly’s lab
group in the Gulf Coast. These are a mitochondrial marker called COI,
which evolves quickly enough that between-species differences can be
viewed, but not so quickly that different members of the same species
cannot be identified as such. Dr. Kelly’s lab is interested in oyster
responses to changing climate, therefore, they are collecting genetic
and physiological data about what those oysters are doing in the
Gulf.</p>
<p>We will be using two R packages today to look at the genetic data of
our oysters:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"haplotype"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_git.html" class="external-link">install_git</a></span><span class="op">(</span><span class="st">"https://gitee.com/zhangrenl/genehapr"</span><span class="op">)</span> </span></code></pre></div>
<p>Our data are stored in a common data format call ‘fasta.’ Let’s click
on a fasta file to open it. You will notice that each file begins with a
line something like</p>
<pre><code>&gt;NU10PO-F_C07_2024-02-23.ab1  (646 bp)</code></pre>
<p>The first two letters in that sequence tell us where the data were
collected. We can cross-reference these with out data spreadsheet from
earlier. The rest tells us some other sequence information, including
how many basepairs we have. This line is called the <code>header</code>
line.</p>
<p>This is followed by the sequence data. fasta is a very commonly-used
format, and many tools are available to work with fasta files. Something
tricky, however, is that our haplotype tools want all our data in one
file. Uh-oh.</p>
<p>First, I’d like you to spitball with a partner about how you might
solve this. No wrong answers, just come up with a solution or two.</p>
<p>Probably, you came up with something like:</p>
<ul>
<li>Read in the files</li>
<li>Get the sequence</li>
<li>Make a list of all the sequences</li>
<li>Write to a file</li>
</ul>
<p>Conceptually quite easy, but needs some knowledge. First, let’s think
about getting a list. For this, we will use a function called
<code>list.files</code>. Have a look at the help
(<code>list.files</code>) and see if you can guess some of the things
you might need to know to make it work.</p>
<p>Roughly, the command is laid out like so:</p>
<pre><code>list_of_files &lt;- list.files(path = Where do you want to look for files? , pattern = What will you look for?)</code></pre>
<p>If we translate to R, this looks like:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">list_of_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"data/"</span>, pattern <span class="op">=</span> <span class="st">".fa"</span><span class="op">)</span></span></code></pre></div>
<p>We will search for the files that end in <code>.fa</code> in the
directory <code>data</code>.</p>
<p>Next, we will need to iteratively read them in. We want all of
them!</p>
<p>We will first work in the <code>geneHapR</code> R package.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">geneHapR</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: ggplot2</span></span></code></pre>
<pre><code><span><span class="co">## Registered S3 method overwritten by 'pegas':</span></span>
<span><span class="co">##   method      from    </span></span>
<span><span class="co">##   [.haplotype genetics</span></span></code></pre>
<p>This package has a tool for reading in the haplotype data Dr. Kelly
collected. But it wants all the sequences in one file. For example, we
can look at one file like so:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/geneHapR/man/import_seqs.html" class="external-link">import_seqs</a></span><span class="op">(</span><span class="st">"../data/GI10PO-F_clean.fa"</span>, format <span class="op">=</span> <span class="st">"fasta"</span><span class="op">)</span></span></code></pre></div>
<p>But we want many files. This is where lists and loops come in. We
have a list of file names. We’re now going to compile these into one big
fasta file. I’m not going to lie to you: this task is utterly obnoxious
to do in R. Just miserable. But it’s very, very easy in UNIX. Let’s take
a moment and go to the UNIX shell. Here is the command we’ll use</p>
<pre><code>cat data/*.fa &gt; combined.fa 
</code></pre>
<p>The UNIX shell is a very old way of interacting with the computer.
It’s fast, efficient, and a little bit terse. <code>cat</code> stands
for concatenate. That means to take one thing and append it to another.
the <code>*</code> is a wildcard. It will match anything. So we would
read this as saying “match anything that ends in .fa, no matter what it
starts with.” The <code>&gt;</code> means to pipe the output into the
combined.fa file. This will contain all our sequences.</p>
<p>Hey real quick - why do you think we’re not writing the combined.fa
file to the data directory right away? Talk about it with a partner for
a minute. Really, there are two reasons. See if you can snag one.</p>
<p>Got it? Let’s read in our file and make sure it worked.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/geneHapR/man/import_seqs.html" class="external-link">import_seqs</a></span><span class="op">(</span><span class="st">"combined.fa"</span>, format <span class="op">=</span> <span class="st">"fasta"</span><span class="op">)</span></span></code></pre></div>
<p>We are going to make a directory called <code>data_processed</code>.
We never want to be in a situation where we might mistake processed data
for raw data, or vice versa. We will then move our file into that
directory.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"data_processed"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="st">"combined.fa"</span>, <span class="st">"data_processed/combined.fa"</span><span class="op">)</span></span></code></pre></div>
<p><code>dir.create</code> makes a directory for a file to go into.
<code>file.copy</code> makes a copy of the file, then moves it into that
directory. There are ways of moving a file without making a copy. Can
anyone guess why it might be advantageous to make a copy first?</p>
<p>Now, load in the copy of the file found in
<code>data_processed</code>. How can you be sure this is the same file
you looked at before? We can look at some properties of the file to
guess at this:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/geneHapR/man/import_seqs.html" class="external-link">import_seqs</a></span><span class="op">(</span><span class="st">"data_processed/combined.fa"</span>, format <span class="op">=</span> <span class="st">"fasta"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">seq2</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">seq</span><span class="op">)</span></span></code></pre></div>
<p>But really, just visually look at it. When you are satisfied that
these are the same file, you can delete the one in the root of your
directory.</p>
<p>Next, our sequences are not the same length because they are
unaligned, meaning we have not discovered what regions are homologous.
We will now do so. The <code>msa</code> package allows us to do a
<code>m</code>ultiple <code>s</code>equence <code>a</code>lignment to
identify these regions.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"msa"</span><span class="op">)</span></span></code></pre></div>
<p>We will align our sequences, then convert back to a format more
commonly used in R:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.bioinf.jku.at/software/msa/" class="external-link">msa</a></span><span class="op">)</span></span>
<span><span class="va">my_alignment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/msa/man/msa.html" class="external-link">msa</a></span><span class="op">(</span><span class="va">seq2</span><span class="op">)</span></span>
<span><span class="va">seq2</span> <span class="op">&lt;-</span>  <span class="fu">msa</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/msa/man/msaConvert.html" class="external-link">msaConvert</a></span><span class="op">(</span><span class="va">my_alignment</span>, <span class="st">"ape::DNAbin"</span><span class="op">)</span></span></code></pre></div>
<p>The sequences are now all one length, and have been aligned. We can
procede with some visuals and mapping. The <code>haplotype</code>
function will find identical sequences between the different
mitochondrial genes from our analysis. These will be the haplotypes. Any
missing data or indels are coded as a <code>?</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"pegas"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://emmanuelparadis.github.io/pegas.html" class="external-link">pegas</a></span><span class="op">)</span></span></code></pre></div>
<p>If you haven’t loaded the above package, please do so now.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">haps</span> <span class="op">&lt;-</span> <span class="fu">pegas</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pegas/man/haplotype.html" class="external-link">haplotype</a></span><span class="op">(</span><span class="va">seq2</span><span class="op">)</span></span></code></pre></div>
<p>We can view the haplotypes, but they aren’t very helpful looking.
Really all you can take away from this view is that most sequences are
at least somewhat unique and different from the others at at least one
site.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">haps</span></span></code></pre></div>
<p>What we’ll do next is put this visually. <code>dist.dna</code> counts
how many differences there are between the haplotypes. For now, we’re
using “N”, or “No Model” and simply tallying distances. You could
however select other models, if you felt it right for your data. The
<code>rmst</code> function computes a network to cover all the different
nodes (haplotypes). Then we plot it.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/dist.dna.html" class="external-link">dist.dna</a></span><span class="op">(</span><span class="va">haps</span>, model <span class="op">=</span> <span class="st">"N"</span><span class="op">)</span></span>
<span><span class="va">nt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/pegas/man/mst.html" class="external-link">rmst</a></span><span class="op">(</span><span class="va">d</span>, B <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>OK, now let’s look at the haplotype network.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">nt</span><span class="op">)</span></span></code></pre></div>
<p>Oh boy, that is something! What is going on here? Let’s look at the
network object.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nt</span></span></code></pre></div>
<p>This tells us that there are 39 haplotypes, some of which are
identical to one another</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pops</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GI"</span>, <span class="st">"GI"</span>, <span class="st">"GI"</span>, <span class="st">"GI"</span>, <span class="st">"GI"</span>, <span class="st">"GI"</span>,<span class="st">"GI"</span>,<span class="st">"GI"</span>, <span class="st">"GI"</span>, <span class="st">"LM"</span>, <span class="st">"LM"</span>, <span class="st">"LM"</span>, <span class="st">"LM"</span>, <span class="st">"LM"</span>, <span class="st">"LM"</span>, <span class="st">"LM"</span>, <span class="st">"LM"</span>, <span class="st">"LM"</span>, <span class="st">"NU"</span>,<span class="st">"NU"</span>, <span class="st">"NU"</span>, <span class="st">"NU"</span>, <span class="st">"NU"</span>, <span class="st">"NU"</span>, <span class="st">"NU"</span>, <span class="st">"NU"</span>, <span class="st">"NU"</span>, <span class="st">"PI"</span>, <span class="st">"PI"</span>, <span class="st">"PI"</span>, <span class="st">"PI"</span>, <span class="st">"PI"</span>, <span class="st">"PI"</span>, <span class="st">"PI"</span>, <span class="st">"PI"</span>, <span class="st">"PI"</span>, <span class="st">"PI"</span>, <span class="st">"PI"</span>, <span class="st">"PI"</span>, <span class="st">"PI"</span> <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">nt</span>, pie <span class="op">=</span> <span class="va">pops</span><span class="op">)</span></span></code></pre></div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by April Wright.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
